lib/exabgp/bgp/message/update/attribute/mprnlri.py

```diff
@@ -69,29 +69,19 @@ def packed_attributes (self, negotiated):
 					# .packed and not .pack()
 					nexthop = nlri.nexthop.ton(negotiated,nlri.afi)
 
-			# mpunli[afi,safi][nexthop] = nlri
-			mpnlri.setdefault((nlri.afi.pack(),nlri.safi.pack()),{}).setdefault(nexthop,[]).append(nlri.pack(negotiated))
-
-		for (pafi,psafi),data in mpnlri.iteritems():
-			for nexthop,nlris in data.iteritems():
-				payload = \
-					pafi + psafi + \
-					chr(len(nexthop)) + nexthop + \
-					chr(0) + ''.join(nlris)
+			# mpunli[nexthop] = nlri
+			mpnlri.setdefault(nexthop,[]).append(nlri.pack(negotiated))
 
-				if self._len(payload) <= maximum:
+		for nexthop,nlris in mpnlri.iteritems():
+			payload = ''.join([self.afi.pack(), self.safi.pack(), chr(len(nexthop)), nexthop, chr(0)])
+			for nlri in nlris:
+				if self._len(payload + nlri) > maximum:
 					yield self._attribute(payload)
+					payload = ''.join([self.afi.pack(), self.safi.pack(), chr(len(nexthop)), nexthop, chr(0), nlri])
 					continue
+				payload  = ''.join([payload, nlri])
 
-				# This will not generate an optimum update size..
-				# we should feedback the maximum on each iteration
-
-				for nlri in nlris:
-					yield self._attribute(
-						pafi + psafi +
-						chr(len(nexthop)) + nexthop +
-						chr(0) + nlri
-					)
+			yield self._attribute(payload)
 
 	def pack (self, negotiated):
 		return ''.join(self.packed_attributes(negotiated))
```

Comment:
So since it was a bug I added a default value for maximum into the code. I'm still not sure that the pack method is every called on these two classes though since I never saw any exceptions or errors during testing that I would have expected had that function ever been called.